### PR TITLE
Add fuzzer for incremental generators

### LINKED ISSUE #68459: Fuzz test the incremental state table
It feels like the [NodeStateTable](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs) tends to have bugs resulting from subtle interactions between steps and across rounds. It may be a good candidate for fuzz testing.

We could write some test code to generate graphs with random arrangements of transforms, operators and so on, random sets of inputs, and then perform random sequences of edits (add, modify, remove of original input).

We could check the output of such tables against a "non-caching" evaluation of the inputs at a given point, or somehow manually construct the outputs we expect based on the operators and transforms in use.

Ideally the code which generates the fuzz-test could also write out a unit test which performs a test of the exact same scenario (i.e. creates the same graph and sequence of inputs which failed during fuzzing).

We have at least one place where we already do fuzz testing, in [TextChangeTests.cs](https://github.com/dotnet/roslyn/blob/826e1770eedb6d8d44a054c9e3ea6569e7ec5970/src/Compilers/Core/CodeAnalysisTest/Text/TextChangeTests.cs#L960), which could be useful as a starting point.

### PR DESCRIPTION
**Pull request by RikkiGibson**
Closes #68459

This fuzzer didn't find any new bugs. It did seem to repro some old bugs like #68451, though.

I'd like to expand this in the future to create more devious generators, but think we should go ahead and get this reviewed as-is.

### CODE REVIEW

**Comment by RikkiGibson** [comment]
As a test-only change we only need 1 sign-off to merge here.

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs [hunk 1]
```diff
@@ -0,0 +1,572 @@
...
+            public string GetNextHintName()
+            {
+                var name = nextHintNameId.ToString();
+                nextHintNameId++;
+                return name;
+            }
+        }
+
+        /// <summary>
+        /// isomorphic to the IncrementalValueProvider tree.
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs:67]:
```suggestion
        /// Isomorphic to the IncrementalValueProvider tree.
```

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+                        """);
+                builder.AppendLine("""
+                                        )
+                        """);
+                builder.AppendLine("""
+                                        .Select((pair, _)
+                                            => (AdditionalText)new InMemoryAdditionalText(
+                                                pair.Left.Path,
+                                                string.Join("", pair.Right.Select(text => text.GetText()!.ToString()))))
+                        """);
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs:239]: Nit: looks like these `AppendLine` calls could be merged together.

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+                        var operator4_1 = makeOperatorTree(@operator, random.Next(depth));
+                        var operator4_2 = makeOperatorTree(@operator, random.Next(depth));
+                        var operator4 = new CombineOperator(operator4_1, operator4_2) { HintNameProvider = hintNameProvider };
+                        return makeOperatorTree(operator4, random.Next(depth));
+                }
+
+                throw ExceptionUtilities.Unreachable();
+            }
+
+            void writeReproducerTest()
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs:418]: Wow, that's cool.

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs [hunk 4]
```diff
@@ -0,0 +1,572 @@
...
+                };
+            }
+
+            void registerPipeline(IncrementalGeneratorInitializationContext context)
+            {
+                // generate a random tree of operations
+                var provider = context.AdditionalTextsProvider;
+
+                // Adjust depth as needed to make a simpler reproducer
+                var depth = random.Next(4);
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs:355]: `depth` looks unused here

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs [hunk 5]
```diff
@@ -0,0 +1,572 @@
...
+                var finalProvider = rootOperator.Apply(provider);
+
+                context.RegisterSourceOutput(finalProvider, (context, text) =>
+                {
+                    context.AddSource(((InMemoryAdditionalText)text).Path, ((InMemoryAdditionalText)text).GetText()!.ToString());
+                });
+            }
+
+            Operator makeOperatorTree(Operator @operator, int depth)
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs:364]: `static`?

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs [hunk 6]
```diff
@@ -0,0 +1,572 @@
...
+                _output.WriteLine(builder.ToString());
+            }
+        }
+
+        [Fact]
+        public void Fuzz_5()
+        {
+            var source = "";
+            var comp = CreateCompilation(source);
+            var hintNameProvider = new HintNameProvider(4);
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs:512]: Didn't understand why this starts at 4 and we don't use `.Reset` on the instance below, but that's probably not important.

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs [hunk 7]
```diff
@@ -0,0 +1,572 @@
...
+                        }
+                    }
+                """);
+
+                _output.WriteLine(builder.ToString());
+            }
+        }
+
+        [Fact]
+        public void Fuzz_5()
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs:508]: What does this test do? Is it an example generated by the test above?

**Review by jjonescz — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs (hunk 7 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs:508]: Yes :)

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs [hunk 8]
```diff
@@ -0,0 +1,572 @@
...
+                };
+            }
+
+            void registerPipeline(IncrementalGeneratorInitializationContext context)
+            {
+                // generate a random tree of operations
+                var provider = context.AdditionalTextsProvider;
+
+                // Adjust depth as needed to make a simpler reproducer
+                var depth = random.Next(4);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs:355]:
```suggestion
```

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs (hunk 6 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs:509]: I think this was generated using an older "version" of the reproducer template. However, I think it's still demonstrating the concept, so would prefer to leave it as is until we have the chance to revisit this in more depth.

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs (hunk 5 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverFuzzTests.cs:361]: This captures 'random' and 'hintNameProvider', so will leave as-is.

**Review by jjonescz — APPROVED** [review]

**Pull request merged**